### PR TITLE
FM-72 & FM-117: Add in module for assets and asset types

### PR DESCRIPTION
**Pull request by leeallen337**
## Why?

https://ndustrialio.atlassian.net/browse/FM-72
https://ndustrialio.atlassian.net/browse/FM-117

Need modules for `Assets` and `Asset Types`

## What changed?
- Added `Assets` module
  - Added methods `create`, `delete`, `get`, `getAll`, `getAllByOrganizationId`, `update`
- Added `AssetTypes` module
  - Added methods `create`, `delete`, `get`, `getAll`, `getAllByOrganizationId`, `update`

### CODE REVIEW

**Comment by leeallen337** [comment]
I haven't bumped the version yet using `npm version`. Probably best to wait until it's been approved and ready to merge

File: src/utils/assets/formatAssetFromServer.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * Normalizes the asset object returned from the API server
+ *
+ * @param {Object} input
+ * @param {string} input.asset_type_id UUID
+ * @param {string} input.created_at ISO 8601 Extended Format date/time string
+ * @param {string} input.description
+ * @param {string} input.id UUID
+ * @param {string} input.label
+ * @param {string} input.organization_id UUID
+ * @param {string} input.updated_at ISO 8601 Extended Format date/time string
```
Comment (greg5green) [inline src/utils/assets/formatAssetFromServer.js]: Looking at this, I realize I need to tighten up the stuff I did last week in #41 in this regard. Which of these is optional?

File: src/utils/assets/formatAssetOptionsToServer.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+/**
+ * Normalizes the options provided when retrieving assets
+ *
+ * @param {Object} options
+ * @param {string} options.assetTypeId ID of the asset type to use for filtering
```
Comment (greg5green) [inline src/utils/assets/formatAssetOptionsToServer.js]: This looks optional

File: src/utils/assets/formatAssetToServer.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+/**
+ * Normalizes the asset object being sent to the API server
+ *
+ * @param {Asset} input
+ *
+ * @returns {Object} output
+ * @returns {string} output.asset_type_id UUID
+ * @returns {string} output.description
```
Comment (greg5green) [inline src/utils/assets/formatAssetToServer.js]: Optional?

File: src/utils/assets/formatAssetTypeFromServer.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
+/**
+ * Normalizes the asset type object returned from the API server
+ *
+ * @param {Object} input
+ * @param {string} input.created_at ISO 8601 Extended Format date/time string
+ * @param {string} input.description
```
Comment (greg5green) [inline src/utils/assets/formatAssetTypeFromServer.js]: Optional?

File: src/utils/assets/formatAssetTypeToServer.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+/**
+ * Normalizes the asset type object being sent to the API server
+ *
+ * @param {AssetType} input
+ *
+ * @returns {Object} output
+ * @returns {string} output.description
```
Comment (greg5green) [inline src/utils/assets/formatAssetTypeToServer.js]: Optional?

File: src/assets/assetTypes.spec.js [hunk 6]
```diff
@@ -0,0 +1,484 @@
...
+      let assetTypeFromServerBeforeFormat;
+      let assetTypeToServerAfterFormat;
+      let assetTypeToServerBeforeFormat;
+      let formatAssetTypeFromServer;
+      let formatAssetTypeToServer;
+      let promise;
+      let request;
+
+      beforeEach(function() {
+        assetTypeFromServerAfterFormat = fixture.build('assetType');
```
Comment (greg5green) [inline src/assets/assetTypes.spec.js]: I like this naming scheme more than what I had been doing 👍

File: src/assets/assetTypes.spec.js [hunk 7]
```diff
@@ -0,0 +1,484 @@
...
+      let request;
+
+      beforeEach(function() {
+        expectedAssetTypeId = fixture.build('assetType').id;
+        assetTypeFromServerAfterFormat = fixture.build('assetType', {
+          id: expectedAssetTypeId
+        });
+        assetTypeFromServerBeforeFormat = fixture.build(
+          'assetType',
+          { id: expectedAssetTypeId },
```
Comment (greg5green) [inline src/assets/assetTypes.spec.js]: You can make this a little simpler actually
```
assetTypeFromServerAfterFormat = fixture.build('assetType');
assetTypeFromServerBeforeFormat = fixture.build('assetType', assetTypeFromServerAfterFormat, { fromServer: true });
```

File: src/assets/assetTypes.spec.js [hunk 8]
```diff
@@ -0,0 +1,484 @@
...
+        );
+      });
+    });
+  });
+
+  describe('getAll', function() {
+    let assetTypesFromServerAfterFormat;
+    let assetTypesFromServerBeforeFormat;
+    let formatAssetTypeFromServer;
+    let numberofAssetTypes;
```
Comment (greg5green) [inline src/assets/assetTypes.spec.js]: `numberOfAssetTypes`?

File: src/assets/assetTypes.spec.js [hunk 9]
```diff
@@ -0,0 +1,484 @@
...
+    beforeEach(function() {
+      numberofAssetTypes = faker.random.number({ min: 1, max: 10 });
+      assetTypesFromServerAfterFormat = fixture.buildList(
+        'assetType',
+        numberofAssetTypes
+      );
+      assetTypesFromServerBeforeFormat = fixture.buildList(
+        'assetType',
+        numberofAssetTypes,
+        { fromServer: true }
```
Comment (greg5green) [inline src/assets/assetTypes.spec.js]: I think there needs to be a `null` here?

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,3 +1,22 @@
...
+  * Assets#update to update an asset
+* Methods around the display and manipulation of Asset Types. They are namespaced under assets (i.e. `assets.types.methodName()`) and include:
+  * AssetTypes#create to add an asset type
+  * AssetTypes#delete to delete an asset type
+  * AssetTypes#get to get an asset type
+  * AssetTypes#getAll to get a list of all asset types
+  * AssetTypes#getAllByOrganizationId to get a list of all asset types for a specific organization
+  * AssetTypes#update to update an asset type
+
 ## [v0.0.16](http://github.com/ndustrialio/contxt-sdk-js/tree/v0.0.15)
```
[the remark this answers is not in the document]
Comment (leeallen337) [reply CHANGELOG.md]: Updated 👍

File: src/assets/index.js [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+import {
+  formatAssetFromServer,
+  formatAssetOptionsToServer,
+  formatAssetToServer
+} from '../utils/assets';
+
+/**
+ * @typedef {Object} Asset
+ * @property {string} assetTypeId UUID corresponding with the asset type
+ * @property {string} description
```
[the remark this answers is not in the document]
Comment (leeallen337) [reply src/assets/index.js]: `description` is optional. Updating 👍

File: src/assets/index.js [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+
+  /**
+   * Get a list of all assets that belong to a particular organization
+   *
+   * API Endpoint: '/organizations/:organizationId/assets'
+   * Method: GET
+   *
+   * @param {string} organizationId UUID corresponding with an organization
+   * @param {Object} options Object containing parameters to be called with the request
+   * @param {string} options.assetTypeId ID of the asset type to use for filtering
```
[the remark this answers is not in the document]
Comment (leeallen337) [reply src/assets/index.js]: They are both optional. Updating 👍

File: src/assets/index.js [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+
+  /**
+   * Updates an asset's data
+   *
+   * API Endpoint: '/assets/:assetId'
+   * Method: PUT
+   *
+   * @param {string} assetId The ID of the asset to update (formatted as a UUID)
+   * @param {Object} update An object containing the updated data for the asset
+   * @param {string} update.description
```
[the remark this answers is not in the document]
Comment (leeallen337) [reply src/assets/index.js]: Yep

File: src/assets/assetTypes.js [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+
+  /**
+   * Updates an asset type's data
+   *
+   * API Endpoint: '/assets/types/:assetTypeId'
+   * Method: PUT
+   *
+   * @param {string} assetTypeId The ID of the asset type to update (formatted as a UUID)
+   * @param {Object} update An object containing the updated data for the asset type
+   * @param {string} update.description
```
[the remark this answers is not in the document]
Comment (leeallen337) [reply src/assets/assetTypes.js]: I don't think we're going to allow them to update `label` for now. See [here](https://github.com/sustainableis/facilities-service/pull/36#discussion_r196870704)

File: src/assets/assetTypes.js [hunk 15]
```diff
@@ -0,0 +1,251 @@
+import isPlainObject from 'lodash.isplainobject';
+import {
+  formatAssetTypeFromServer,
+  formatAssetTypeToServer
+} from '../utils/assets';
+
+/**
+ * @typedef {Object} AssetType
+ * @property {string} description
```
[the remark this answers is not in the document]
Comment (leeallen337) [reply src/assets/assetTypes.js]: Currently. `description` for an `Asset` is optional but it is required for an `AssetType`. See the [required fields](https://github.com/sustainableis/facilities-service/blob/master/controllers/asset_types.js#L12). I think all fields are required for `AssetType`

File: src/utils/assets/formatAssetFromServer.js (hunk 1 above)
Comment (leeallen337) [reply src/utils/assets/formatAssetFromServer.js]: `input.description` is optional. Updating 👍

File: src/utils/assets/formatAssetToServer.js (hunk 3 above)
Comment (leeallen337) [reply src/utils/assets/formatAssetToServer.js]: Yep

File: src/utils/assets/formatAssetTypeToServer.js (hunk 5 above)
Comment (leeallen337) [reply src/utils/assets/formatAssetTypeToServer.js]: Not for `AssetType`

File: src/assets/assetTypes.js [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+   *   .getAll()
+   *   .then((assetTypes) => console.log(assetTypes))
+   *   .catch((err) => console.log(err));
+   */
+  getAll() {
+    return this._request
+      .get(`${this._baseUrl}/assets/types`)
+      .then((response) => {
+        return {
+          ...response,
```
[the remark this answers is not in the document]
Comment (leeallen337) [reply src/assets/assetTypes.js]: Updated with a formatter

File: src/assets/assetTypes.js [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+  }
+
+  /**
+   * Gets a list of all asset types
+   *
+   * API Endpoint: '/assets/types/
+   * Method: GET
+   *
+   * @returns {Promise}
+   * @fulfill {AssetType[]} Information about all asset types
```
[the remark this answers is not in the document]
Comment (leeallen337) [reply src/assets/assetTypes.js]: Updated with a type

File: src/assets/assetTypes.js [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+          "An organization ID is required for getting a list of an organization's asset types."
+        )
+      );
+    }
+
+    return this._request
+      .get(`${this._baseUrl}/organizations/${organizationId}/assets/types`)
+      .then((response) => {
+        return {
+          ...response,
```
[the remark this answers is not in the document]
Comment (leeallen337) [reply src/assets/assetTypes.js]: Updated with a formatter

File: src/assets/assetTypes.js [hunk 19]
```diff
@@ -0,0 +1,251 @@
...
+  /**
+   * Gets a list of all asset types that belong to a particular organization
+   *
+   * API Endpoint: '/organizations/:organizationId/assets/types'
+   * Method: GET
+   *
+   * @param {string} organizationId UUID corresponding with an organization
+   *
+   * @returns {Promise}
+   * @fulfill {AssetType[]} Information about all asset types
```
[the remark this answers is not in the document]
Comment (leeallen337) [reply src/assets/assetTypes.js]: Updated with a type

File: src/assets/index.js [hunk 20]
```diff
@@ -0,0 +1,257 @@
...
+  }
+
+  /**
+   * Get a list of all assets
+   *
+   * API Endpoint: '/assets'
+   * Method: GET
+   *
+   * @returns {Promise}
+   * @fulfill {Asset[]} Information about all assets
```
[the remark this answers is not in the document]
Comment (leeallen337) [reply src/assets/index.js]: Updated with a type

File: src/assets/index.js [hunk 21]
```diff
@@ -0,0 +1,257 @@
...
+   * @example
+   * contxtSdk.assets
+   *   .getAll()
+   *   .then((assets) => console.log(assets))
+   *   .catch(err) => console.log(err);
+   */
+  getAll() {
+    return this._request.get(`${this._baseUrl}/assets`).then((response) => {
+      return {
+        ...response,
```
[the remark this answers is not in the document]
Comment (leeallen337) [reply src/assets/index.js]: Updated with a formatter

File: src/assets/index.js [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+
+    const params = formatAssetOptionsToServer(options);
+
+    return this._request
+      .get(`${this._baseUrl}/organizations/${organizationId}/assets`, {
+        params
+      })
+      .then((response) => {
+        return {
+          ...response,
```
[the remark this answers is not in the document]
Comment (leeallen337) [reply src/assets/index.js]: Updated with a formatter

**Comment by leeallen337** [comment]
@greg5green I believe I've addressed all your comments. I added in a formatter for the pagination data and some new specs and updates the way the `getAll` and `getAllByOrganizationId` methods were being handled. Whenever you have time it could use another review

**Pull request merged**